src/compiler/checker.ts

```diff
@@ -7504,12 +7512,18 @@ namespace ts {
                 // will never be observed because a qualified name can't reference signatures.
                 if (symbol.flags & (SymbolFlags.Function | SymbolFlags.Method)) {
                     type.callSignatures = getSignaturesOfSymbol(symbol);
-                    type.constructSignatures = filter(type.callSignatures, sig => isJSConstructor(sig.declaration));

```

Comment:
Construct signatures are now constructed just below here, for symbols marked `Class | Function` only, instead of running this on all functions.